Notify correct timeout in message

ubergesundheit opened the pull request:
I plan to use nanny to monitor processes that rune once a day, so my I guess it should be possible to:

- Notify nanny upon start of the process (say at 8am) with next_signal set to `10m` because I know if the process does not finish in 10 minutes, there is something wrong
- Notify nanny upon end of the process (8:07 am) with next_signal set to `23h53m10s` to alert if the process hasn't been started at 8am next day

But nanny only reports the next signal of the very first occurence of the signal

This pull request introduces a new struct which wraps a time.Timer together with its signal to be able to have the most recent NextSignal value available for messaging.

ubergesundheit commented:
Oh! How can I reproduce the data race? If I can reproduce, I am happy to fix it!

I ran all the `make run,test,vet,lint` commands from the contributing section in the readme.

lunemec commented:
Yes, there is no test for this. I ran into it by curling it a few times.
But basically it is caused by missing mutex lock when accessing the shared structure `nannyTimer` from several goroutines.

I managed to produce it with these:
```bash
curl -v http://localhost:8080/api/v1/signal --data '{ "name": "", "notifier": "stderr", "next_signal": "1s", "meta": {"metakey": "metavalue"}}'
curl -v http://localhost:8080/api/v1/signal --data '{ "name": "", "notifier": "stderr", "next_signal": "5s", "meta": {"metakey": "metavalue"}}'
```

But easiest would be to create a test, where you'd hit it from some hundreds of goroutines with varying next_signal. There is already one test similar to that which you can use as template (`TestConcurrent`).

ubergesundheit commented:
Thanks for the pointers! I did the test setup analog to `TestConcurrent` with two additional loops inside goroutines which call handle with different next_signal values but I couldn't provoke the data race.

Maybe you could help me with some additional pointers or tell me how the structure of the test could provoke the behaviour?

lunemec commented:
Hmm, that is strange it didn't trigger the data race, however basically 2 goroutines should do the trick, but you basically need to have one that will call
```
  nanny/pkg/nanny.(*nannyTimer).Reset()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/nanny.go:56 +0x45
```
- the write part

and another goroutine which is sleeping (previous timer) which is calling:
```
  nanny/pkg/nanny.(*nannyTimer).Reset()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/nanny.go:56 +0x45
```

Since when you create the timer, it spawns it in the goroutine. So the struct (`timerStruct`) you are passing to the goroutine will be shared between that timer goroutine, and maybe another goroutine spawned by the http server, that will be calling `timer.Reset()`.

I'm also thinking if it would not be better to create interface from this, where the methods would properly call Lock() and Unlock() when reading/writing the fields. But it is definitely doable without interfaces...

ubergesundheit commented:
What do you think about

- Changing the signature of `nannyTimer.Reset` to receive just a `time.Duration` which is then changed in the signal of the `nannyTimer` like `n.signal.NextSignal = newNextSignal`

or:

- Changing the type of the signal in the `nannyTimer` struct to `*validSignal`?

Anyway here is my test:
```golang
func TestConcurrentMsgChange(t *testing.T) {
	n := nanny.Nanny{Name: "test msg changing nanny"}
	dummy := &DummyNotifier{}
	signal := nanny.Signal{
		Name:       "test msg changing program",
		Notifier:   dummy,
		NextSignal: time.Duration(1) * time.Second,
	}

	err := n.Handle(signal)
	if err != nil {
		t.Errorf("n.Signal should not return error, got: %v\n", err)
	}

	// Before the `NextSignal` duration, nothing should happen.
	dummyMsg := dummy.NotifyMsg()
	if dummyMsg.Program != "" {
		t.Errorf("dummy msg should be empty before NextSignal time expires: %v\n", dummyMsg)
	}

	go func() {
		for i := 0; i < 10000000000; i++ {
			err := n.Handle(nanny.Signal{
				Name:       "test msg changing program",
				Notifier:   dummy,
				NextSignal: time.Duration(500) * time.Millisecond,
			})
			if err != nil {
				t.Errorf("n.Signal should not return error, got: %v\n", err)
			}
		}
	}()
	go func() {
		for i := 0; i < 10000000000; i++ {
			err := n.Handle(nanny.Signal{
				Name:       "test msg changing program",
				Notifier:   dummy,
				NextSignal: time.Duration(800) * time.Millisecond,
			})
			if err != nil {
				t.Errorf("n.Signal should not return error, got: %v\n", err)
			}
		}
	}()

	// Call handle with different nextsignal again to simulate program calling before notification.
	err = n.Handle(nanny.Signal{
		Name:       "test msg changing program",
		Notifier:   dummy,
		NextSignal: time.Duration(1) * time.Second,
	})
	if err != nil {
		t.Errorf("n.Signal should not return error, got: %v\n", err)
	}
	// After 1s, DummyNotifier should return error.
	time.Sleep(time.Duration(1)*time.Second + time.Duration(100)*time.Millisecond)
	dummyMsg = dummy.NotifyMsg()
	if dummyMsg.Program == "" {
		t.Errorf("dummy msg should not be empty after NextSignal time expired: %v\n", dummyMsg)
	}
}
```

lunemec commented:
`nannyTimer.Reset` having signature identical with https://golang.org/pkg/time/#Timer.Reset would be a good idea.

And why would you want `validSignal` to be pointer? Does it fix the data race, or I'm missing something?

lunemec commented on pkg/nanny/nanny.go:
```diff
@@ -97,7 +97,7 @@ func (n *Nanny) handle(s validSignal) error {
 	if timer != nil {
 		// Timer exists, reset the timer to the new signal value.
 		n.lock.Lock()
```
Can you please move this lock/unlock to the `Reset` function?

ubergesundheit commented:
Ok, I changed the signature and also changed the check for the CallbackFunc to use the CallbackFunc from the nannyTimers signal stored in the hashmap.

For the idea with the pointers: I don't know if it makes a difference, because I am still not able to reproduce the data race. I will try to come up with a test that produces the data race!

lunemec commented:
I managed to trigger the data race manually again, I'll try to create test for it.

lunemec commented:
This test will consistently trigger the data race, you only have to provide `CallbackFunc`:
```go
func TestNannyTimer(t *testing.T) {
	n := nanny.Nanny{Name: "test msg changing nanny"}
	dummy := &DummyNotifier{}
	signal := nanny.Signal{
		Name:         "test msg changing program",
		Notifier:     dummy,
		NextSignal:   time.Duration(2) * time.Second,
		CallbackFunc: func(s *nanny.Signal) {},
	}
	err := n.Handle(signal)
	if err != nil {
		t.Errorf("n.Signal should not return error, got: %v\n", err)
	}

	// Trigger the first signal's error.
	time.Sleep(time.Duration(2) * time.Second)

	// Before the `NextSignal` duration, nothing should happen.
	dummyMsg := dummy.NotifyMsg()
	if dummyMsg.Program == "" {
		t.Errorf("dummy msg should not be empty after NextSignal time expires: %v\n", dummyMsg)
	}
	// Call handle with different nextsignal again to simulate program calling before notification.
	err = n.Handle(nanny.Signal{
		Name:         "test msg changing program",
		Notifier:     dummy,
		NextSignal:   time.Duration(1) * time.Second,
		CallbackFunc: func(s *nanny.Signal) {},
	})
	if err != nil {
		t.Errorf("n.Signal should not return error, got: %v\n", err)
	}
	// After 1s, DummyNotifier should return error.
	time.Sleep(time.Duration(1)*time.Second + time.Duration(100)*time.Millisecond)
	dummyMsg = dummy.NotifyMsg()
	if dummyMsg.Program == "" {
		t.Errorf("dummy msg should not be empty after NextSignal time expired: %v\n", dummyMsg)
	}

	msg := dummyMsg.Format()
	if strings.Contains(msg, "2s") {
		t.Errorf("dummy msg should not contain 1s after NextSignal time expired: %v\n", dummyMsg)
	}
}
```

lunemec commented:
Thanks, however tests do not pass and there is another data race (different one):
```
==================
WARNING: DATA RACE
Write at 0x00c4203d21b0 by goroutine 6:
  nanny/pkg/nanny.(*nannyTimer).Reset()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/timer.go:28 +0x87
  nanny/pkg/nanny.(*Nanny).handle()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/nanny.go:87 +0xaa
  nanny/pkg/nanny.(*Nanny).Handle()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/nanny.go:62 +0x235
  nanny/pkg/nanny_test.TestNannyTimer()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/nanny_test.go:305 +0x38c
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:777 +0x16d

Previous read at 0x00c4203d21b0 by goroutine 9:
  nanny/pkg/nanny.(*nannyTimer).onExpire()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/timer.go:47 +0x131
  nanny/pkg/nanny.(*nannyTimer).(nanny/pkg/nanny.onExpire)-fm()
      /home/lulu/Documents/git/Go/src/nanny/pkg/nanny/timer.go:21 +0x41

Goroutine 6 (running) created at:
  testing.(*T).Run()
      /usr/lib/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/lib/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/lib/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/lib/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/lib/go/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:108 +0x324

Goroutine 9 (finished) created at:
  time.goFunc()
      /usr/lib/go/src/time/sleep.go:172 +0x51
==================
```

lunemec commented:
I believe you can fix the data race by moving the mutex from `Nanny` struct to the `nannyTimer` struct.

ubergesundheit commented:
You were right 🎉  I thought I tried that, but I tried again and it worked!

There is one warning from `make lint` though:
    pkg/nanny/nanny.go:95:39:warning: exported method GetTimer returns unexported type *nanny.nannyTimer, which can be annoying to use (golint)

Should I fix it?

ubergesundheit commented:
mh the tests don't fail locally on my machine (mac, go version go1.10.3 darwin/amd64) :(

Cloned it again from my branch to see if I missed something to commit, reran the tests but It still works. 

I'll now try in a docker container to see If I can produce the travis error

lunemec commented:
Yes, they do not fail on my machine either, I'll manually restart travis.

ubergesundheit commented:
Reproduced it in a container:
```
root@24fa49d46bc3:/go/src/nanny# make test
go test -race -cover -v ./...
?   	nanny	[no test files]
=== RUN   TestAPISignal
--- PASS: TestAPISignal (0.00s)
=== RUN   TestPersistence
--- PASS: TestPersistence (0.00s)
PASS
coverage: 0.0% of statements
ok  	nanny/api	1.075s	coverage: 0.0% of statements
?   	nanny/cmd	[no test files]
=== RUN   TestNanny
--- PASS: TestNanny (1.10s)
=== RUN   TestNannyDoesNotNotify
--- PASS: TestNannyDoesNotNotify (0.90s)
=== RUN   TestEmptyNanny
--- PASS: TestEmptyNanny (0.00s)
=== RUN   TestNannyCallsErrorFunc
--- PASS: TestNannyCallsErrorFunc (1.10s)
=== RUN   TestNextSignalZero
--- PASS: TestNextSignalZero (0.00s)
=== RUN   TestConcurrent
--- PASS: TestConcurrent (1.10s)
=== RUN   TestMultipleTimerResets
--- PASS: TestMultipleTimerResets (1.11s)
=== RUN   TestMsgChange
--- PASS: TestMsgChange (1.10s)
=== RUN   TestNannyTimer
==================
WARNING: DATA RACE
Read at 0x00c420374390 by goroutine 74:
  nanny/pkg/nanny.(*nannyTimer).notify()
      /go/src/nanny/pkg/nanny/timer.go:65 +0x136
  nanny/pkg/nanny.(*nannyTimer).onExpire()
      /go/src/nanny/pkg/nanny/timer.go:36 +0x68
  nanny/pkg/nanny.(*nannyTimer).(nanny/pkg/nanny.onExpire)-fm()
      /go/src/nanny/pkg/nanny/timer.go:22 +0x41

Previous write at 0x00c420374390 by goroutine 6:
  nanny/pkg/nanny.(*nannyTimer).Reset()
      /go/src/nanny/pkg/nanny/timer.go:31 +0x97
  nanny/pkg/nanny.(*Nanny).handle()
      /go/src/nanny/pkg/nanny/nanny.go:84 +0xaa
  nanny/pkg/nanny.(*Nanny).Handle()
      /go/src/nanny/pkg/nanny/nanny.go:59 +0x235
  nanny/pkg/nanny_test.TestNannyTimer()
      /go/src/nanny/pkg/nanny/nanny_test.go:305 +0x37e
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d

Goroutine 74 (running) created at:
  time.goFunc()
      /usr/local/go/src/time/sleep.go:172 +0x51

Goroutine 6 (running) created at:
  testing.(*T).Run()
      /usr/local/go/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      /usr/local/go/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      /usr/local/go/src/testing/testing.go:777 +0x16d
  testing.runTests()
      /usr/local/go/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      /usr/local/go/src/testing/testing.go:978 +0x2cd
  main.main()
      _testmain.go:108 +0x324
==================
--- FAIL: TestNannyTimer (3.10s)
	nanny_test.go:302: dummy msg should not be empty after NextSignal time expires: {  0s map[]}
	testing.go:730: race detected during execution of test
FAIL
coverage: 95.3% of statements
FAIL	nanny/pkg/nanny	9.604s
?   	nanny/pkg/notifier	[no test files]
=== RUN   TestSQLiteDataConsistency
--- PASS: TestSQLiteDataConsistency (0.00s)
PASS
coverage: 55.6% of statements
ok  	nanny/pkg/storage	1.017s	coverage: 55.6% of statements
?   	nanny/pkg/version	[no test files]
Makefile:28: recipe for target 'test' failed
make: *** [test] Error 1
```

ubergesundheit commented:
But only once :(
As you've already guessed, I am really bad with go concurrency..

lunemec commented:
Yes it is sometimes difficult. In this case, you have 2 goroutines, 1 is the timer in the background, which was just calling the `onExpire()` func and the other was incomming `handle()` call which caused `nannyTimer.Reset()` being called. 

`onExpire()` then called `notify()` which reads shared `nannyTimer` struct, and the `timer.Reset` tries to write into `nannyTimer` struct.

There are 2 solutions.
1) add locks to `notify()`, ideally remove them where they are not needed, and just lock the parts where the shared struct items are read/wrote.
2) use chan's, but that might be very strange here.

So, using the locks, you need 1 lock in the `Reset()` function, another in `onExpire` in the callback part, but please don't use defer on this one, because it might hold the lock longer than it actually needs to. And final lock would be in `notify` function, where you read from nt.signal (which might change in Restet).

lunemec commented on pkg/nanny/timer.go:
```diff
@@ -0,0 +1,70 @@
+func newNannyTimer(s validSignal, nanny *Nanny) *nannyTimer {
+	timer := &nannyTimer{signal: s, nanny: nanny}
+	timer.timer = time.AfterFunc(timer.signal.NextSignal, timer.onExpire)
+	return timer
```
This could be:
```go
return &timer{
    signal: s,
    nanny: nanny,
    timer: time.AfterFunc(timer.signal.NextSignal, timer.onExpire),
}
```

lunemec replied:
Oh sorry, it could not.

lunemec commented:
Do you want to make more changes to it, or can I merge?

ubergesundheit commented:
I'll rename the `nannyTimer` struct to `Timer` as per your suggestion.

After that, I think its up to you to decide..

ubergesundheit commented:
thanks!
